code/obj/item/paper.dm

```diff
@@ -173,49 +186,168 @@
+			else
+				boutput(usr, "There is no where else you can stamp!")
+			. = TRUE
 
-	if (href_list["form"])
-		if (istype(usr.equipped(), /obj/item/pen))
-			. = href_list["form"]
-			if (. in form_startpoints)
-				// fill in field (up to field length)
-				var/t = input(usr, "What text do you wish to add?", "[src.name]", null) as null|text
-				if (!t)
-					return
-				if ((!in_range(src, usr) && src.loc != usr && !( istype(src.loc, /obj/item/clipboard) ) && src.loc.loc != usr))
-					return
-				t = copytext(html_encode(t), 1, (form_endpoints[.] - form_startpoints[.]) + 1)
-				t = sign_name(t, usr)
-				src.info = copytext(src.info, 1, form_startpoints[.]) + "" + t + "" + copytext(src.info, form_startpoints[.] + length(t))
-				build_formpoints()
-				usr.examine_verb(src)
-
-		if (istype(usr.equipped(), /obj/item/stamp))
-			var/obj/item/stamp/S = usr.equipped()
-			. = href_list["form"]
-			if (S && (. in form_startpoints))
-				switch(input(usr, "Stamp \the [src] with [S.name]?", "[src.name]", "No") in list("Yes", "No"))
-					if ("No")
-						return
-					if ("Yes")
-						var/T = S.get_stamp_text()
-						if (!T)
-							return
-						if ((!in_range(src, usr) && src.loc != usr && !(istype(src.loc, /obj/item/clipboard)) && src.loc.loc != usr))
-							return
-						// replace whole form field with stamp
-						src.info = copytext(src.info, 1, form_startpoints[.]) + "" + T + "" + copytext(src.info, form_endpoints[.])
-						src.icon_state = "paper_stamped"
-						build_formpoints()
-						usr.examine_verb(src)
-
-	src.add_fingerprint(usr)
-
-/obj/item/paper/attackby(obj/item/P as obj, mob/user as mob)
+		if("save")
+			if (src.icon_state == "paper_blank" && params["text"])
+				src.icon_state = "paper"
+			var/in_paper = params["text"]
+			var/paper_len = length(in_paper)
+			field_counter = params["field_counter"] ? text2num(params["field_counter"]) : field_counter
+
+			if(paper_len > MAX_PAPER_LENGTH)
+				// Side note, the only way we should get here is if
+				// the javascript was modified, somehow, outside of
+				// byond.  but right now we are logging it as
+				// the generated html might get beyond this limit
+				logTheThing("PAPER: [key_name(ui.user)] writing to paper [name], and overwrote it by [paper_len-MAX_PAPER_LENGTH]")
+			if(paper_len == 0)
+				boutput(ui.user, pick("Writing block strikes again!", "You forgot to write anthing!"))
+			else
+				logTheThing("PAPER: [key_name(ui.user)] writing to paper [name]")
+				if(info != in_paper)
+					boutput(ui.user, "You have added to your paper masterpiece!");
+					info = in_paper
+					update_static_data(usr,ui)
+			. = TRUE
+
+/obj/item/paper/ui_static_data(mob/user)
+	. = list()
+	.["name"] = src.name
+	.["sizeX"] = src.sizex
+	.["sizeY"] = src.sizey
+	.["text"] = src.info
+	.["max_length"] = MAX_PAPER_LENGTH
+	.["paper_color"] = (color || "white")	// color might not be set
+	.["paper_state"] = src.icon_state	/// TODO: show the sheet will bloodied or crinkling?
```

Comment:
todo?

Reply:
Was going to do some stuff for singed paper but didn't, removed the line.